x-pack/plugins/lens/public/text_based_languages_datasource/text_based_languages.tsx

```diff
@@ -0,0 +1,586 @@
+          [newLayerId]: blankLayer(index, query, columns),
+        },
+      };
+    },
+    createEmptyLayer() {
+      return {
+        indexPatternRefs: [],
+        layers: {},
+        fieldList: [],
+      };
+    },
+
+    cloneLayer(state, layerId, newLayerId, getNewId) {
+      return {
+        ...state,
+      };
+    },
+
+    removeLayer(state: TextBasedLanguagesPrivateState, layerId: string) {
+      const newLayers = {
+        ...state.layers,
+        [layerId]: {
+          ...state.layers[layerId],
+          columns: [],
+        },
+      };
+
+      return {
+        ...state,
+        layers: newLayers,
+        fieldList: state.fieldList,
+      };
+    },
+
+    clearLayer(state: TextBasedLanguagesPrivateState, layerId: string) {
+      return {
+        ...state,
+        layers: {
+          ...state.layers,
+          [layerId]: { ...state.layers[layerId], columns: [] },
+        },
+      };
+    },
+
+    getLayers(state: TextBasedLanguagesPrivateState) {
+      return state && state.layers ? Object.keys(state?.layers) : [];
+    },
+    getCurrentIndexPatternId(state: TextBasedLanguagesPrivateState) {
+      const layers = Object.values(state.layers);
+      return layers?.[0]?.index;
+    },
+    isTimeBased: (state, indexPatterns) => {
+      if (!state) return false;
+      const { layers } = state;
+      return (
+        Boolean(layers) &&
+        Object.values(layers).some((layer) => {
+          return Boolean(indexPatterns[layer.index]?.timeFieldName);
+        })
+      );
+    },
+    getUsedDataView: (state: TextBasedLanguagesPrivateState, layerId: string) => {
+      return state.layers[layerId].index;
+    },
+
+    removeColumn({ prevState, layerId, columnId }) {
+      return {
+        ...prevState,
+        layers: {
+          ...prevState.layers,
+          [layerId]: {
+            ...prevState.layers[layerId],
+            columns: prevState.layers[layerId].columns.filter((col) => col.columnId !== columnId),
+          },
+        },
+      };
+    },
+
+    toExpression: (state, layerId, indexPatterns) => {
+      return toExpression(state, layerId);
+    },
+
+    renderDataPanel(
+      domElement: Element,
+      props: DatasourceDataPanelProps<TextBasedLanguagesPrivateState>
+    ) {
+      render(
+        <I18nProvider>
+          <TextBasedLanguagesDataPanel
+            data={data}
+            dataViews={dataViews}
+            expressions={expressions}
+            {...props}
+          />
+        </I18nProvider>,
+        domElement
+      );
+    },
+
+    renderDimensionTrigger: (
+      domElement: Element,
+      props: DatasourceDimensionTriggerProps<TextBasedLanguagesPrivateState>
+    ) => {
+      const columnLabelMap = TextBasedLanguagesDatasource.uniqueLabels(props.state);
+      const layer = props.state.layers[props.layerId];
+      const selectedField = layer?.allColumns?.find(
+        (column) => column.columnId === props.columnId
+      )!;
```

Comment:
the selected field might be not available, we should handle this case by rendering a "FIELD MISSING" label or something like that.

Reply:
Done https://github.com/elastic/kibana/pull/140469/commits/309b68c5b4c1dc590788dd0e98cd15b39dd12e0d